### ISSUE TITLE
remove AMP 64k parameter limit

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#9488]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#9488|
|Type|enhancement|
|Created|2018-07-13 03:26:31Z|

I have now had practical experience with AMP at 4 different companies, and every time I have deployed it in real life, we've had a couple of cases where we just need to jam a megabyte of data into a parameter.  The 64k limit has never served its role as an inducement to break data up into more manageable chunks, or to come up with some kind of streaming system to parcel out data incrementally; it's always been a massive inconvenience that we hit only after having successfully tested with smaller payloads.  The data is never originating within the AMP messaging system, it's always coming from somewhere else, so we are already generally dealing with large message bodies.

This shouldn't even be a special parameter type or anything; `amp.Argument` should do the chunking and unchunking itself.

<details><summary>Searchable metadata</summary>

```
trac-id__9488 9488
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__None None
branch__ 
branch_author__ 
status__new new
resolution__None None
component__core core
keywords__None None
time__1531452391298698 1531452391298698
changetime__1603256863519369 1603256863519369
version__None None
owner__None None
cc__teratorn cc__altendky
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)<a name="note_1"></a>|@teratorn commented|
|-|-|

I think there are a few things to consider when looking at "fixing" AMP

A) compatibility between existing AMPv1 peers and AMPv2.. whatever AMPv2 turns out to be, say:
   * extending length limit with key.1, key.2, key.3 etc
   * keeping length-limited ListOf
   * maybe adding special cases of AmpList to handle unlimited-length data structures
   * protocol version negotiation

B) admitting that the AMPv1 encoding scheme is fundamentally limited and can't both support known-length coding and unlimited-length streaming easily, and we could do better by using netstrings everywhere, making implementation easier and extending that to support streaming.... which sounds fairly promising to me.

C) We might solve the problem with AMPv2, of not knowing the encoded length ahead of time, by using key names in reverse order, e.g. key.5, key.4.. ordering would indicate that only N+1 keys will follow.. e.g. if the first key seen is key.1 it would mean 1+1 keys are expected in total: key.1 and key, in that order. Data would still be coded in the normal order.. just the key naming scheme would be inverted to indicate the number of keys used to encode the known-length data.
Then you would know roughly, within 64k, how long the data was ahead of time.
If on the other hand you want to stream data of an unknown-length you would encode the key names in the normal order, key, key.1, key.2, which can encode data of any length.

I think ideally an application wants to write as much data as feasible to a stream socket at once, with minimal chunking.. so AMP ought to support arbitrary fixed-length encoding, and variable-length streaming.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

There are 2 broad use-cases for AMP that I can think of:

- Internal communication in a system where no version negotiation is necessary
- Communication in a system over the public Internet, where AMP will be frontend by TLS

Given that we always have a TLS wrapper for #2, perhaps we could use ALPN for the wire-syntax negotiation phase?  This sidesteps almost all the issues above, but I think still satisfies the two use-cases pretty clearly.

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)<a name="note_4"></a>|@teratorn commented|
|-|-|

Replying to [Glyph](#note_2):
> There are 2 broad use-cases for AMP that I can think of:
> 
> - Internal communication in a system where no version negotiation is necessary

Ok

> - Communication in a system over the public Internet, where AMP will be frontend by TLS
> 
> Given that we always have a TLS wrapper for #2, perhaps we could use ALPN for the wire-syntax negotiation phase?  This sidesteps almost all the issues above, but I think still satisfies the two use-cases pretty clearly.

Ok, could you please be specific about which problems this side-steps? It sounds good to me, though.

What's the next step?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

Replying to [teratorn](#note_4):
> Replying to [Glyph](#note_2):
> > There are 2 broad use-cases for AMP that I can think of:
> > 
> > - Internal communication in a system where no version negotiation is necessary
> 
> Ok
> 
> > - Communication in a system over the public Internet, where AMP will be frontend by TLS
> > 
> > Given that we always have a TLS wrapper for #2, perhaps we could use ALPN for the wire-syntax negotiation phase?  This sidesteps almost all the issues above, but I think still satisfies the two use-cases pretty clearly.
> 
> Ok, could you please be specific about which problems this side-steps? It sounds good to me, though.

We don't have to consider A — i.e. compatibility between AMPv1 and AMPv2 — at the wire level, because it would either be communicated implicitly by configuration (the process pool case) or out of band by ALPN (the on-the-Internet case).  Ideally we'd still want compatibility at the API level (in all its various implementations)so it would have to be semantically quite similar.

This means we can just use netstrings, and also avoid considering the complexities involved in C.

> What's the next step?

Let's get a replacement for BinaryBoxProtocol going, one that uses netstrings.

The next step after that is to do an ALPN thing so that the `AMP` class can automatically be smart.  This is the `acceptableProtocols` parameter to CertificateOptions and optionsForClientTLS, as well as the https://twistedmatrix.com/documents/18.7.0/api/twisted.internet.interfaces.INegotiated.html#negotiatedProtocol attribute.  (I think there might be a ticket somewhere for letting a protocol factory select its own acceptable protocols, which would let AMP do this more directly when listening over TLS, but that's a separate issue.)

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_6"></a>|@rodrigc set owner to @rodrigc<br>@rodrigc set status to `closed`|
|-|-|

In changeset f6980e302bf084acdabe0eaf03f43053b0e73f94
```
#!CommitTicketReference repository="" revision="f6980e302bf084acdabe0eaf03f43053b0e73f94"
Merge pull request #1417 from zerospam/2529-jothan-ampv2

Author: jothan
Reviewer: rodrigc
Fixes: ticket:2529 ticket:9488

Implement AMP version 2
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

In changeset 8057ddc2d6cd7d318e1e4fc239c8726c1e1dbc28
```
#!CommitTicketReference repository="" revision="8057ddc2d6cd7d318e1e4fc239c8726c1e1dbc28"
Merge pull request #1455 from twisted/revert-1417-2529-jothan-ampv2

Revert "Implement AMP version 2"

As discussed over at https://twistedmatrix.com/pipermail/twisted-python/2020-October/065288.html

Reopens: ticket:9488
Reopens: ticket:2529
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set status to `reopened`|
|-|-|

I guess "reopens" doesn't work right any more. Hm.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph removed owner<br>@glyph set status to `new`|
|-|-|
